keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,540 @@
+---
+title: Object Bucket Provisioning
+authors:
+  - "@jeffvance"
+  - "@copejon"
+owning-sig: sig-storage
+participating-sigs:
+  - sig-storage
+reviewers:
+  - "@saad-ali"
+  - "@alarge"
+  - "@erinboyd"
+  - "@guymguym"
+  - "@travisn"
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2019-11-25
+last-updated: 2019-11-15
+status: implementable
+see-also:
+  - "https://github.com/kube-object-storage/lib-bucket-provisioner"
+  - "https://github.com/kube-object-storage/lib-bucket-provisioner/blob/master/doc/design/object-bucket-lib.md"
+  - "https://github.com/kube-object-storage/lib-bucket-provisioner/tree/master/doc/examples"
+---
+
+# Object Bucket Provisioning
+
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+- [Design Details](#design-details)
+  - [Custom Resources](#custom-resources)
+  - [Moving Parts](#moving-parts)
+  - [Interfaces](#interfaces)
+  - [Library Usage](#library-usage)
+  - [Current Restrictions](#current-restrictions)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+    - [Examples](#examples)
+      - [Alpha -&gt; Beta Graduation](#alpha---beta)
+      - [Beta -&gt; GA Graduation](#beta---ga)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Implementation History](#implementation-history)
+- [Drawbacks [optional]](#drawbacks-optional)
+- [Alternatives [optional]](#alternatives-optional)
+<!-- /toc -->
+
+## Summary
+
+Kubernetes natively supports dynamic provisioning for file and block storage but lacks support for object bucket provisioning.
+Without a formal API to describe and manage bucket provisioning, storage providers are left to write their own provisioners including all of the necessary controller code.
```

Comment:
I'm uncomfortable with the implied idea that object bucket provisioning is just an extension of the existing block and file support.

Block and file support are accessed from within containers using local standard APIs.  Object storage is logically just a remote service - there's nothing in particular that makes it inherently similar to block/file.  Imagine that you wanted to be able to support provisioning against various "relational database as a service" products (e.g., RDS, CloudSQL, etc.).  You could imagine an api for things like "create a database" and "create credentials for accessing this database".  And you'd interact with the database via some sort of SQL client.  I see such a system as a much closer analog for object storage than block/file.

Reply:
Fair point. When I think of storage I think of file, block and object. I tend to lump object with file and block more than with databases because of its simplicity and narrower scope. 
Of course the fact that there is no mount for object makes it similar, from a pod's perspective, to a database. 
From a provisioner author's perspective I suppose there's no real difference between a bucket API and a DB API.

Reply:
I'll leave some time for others to chime in and then will push changes to this doc to reflect your comments.

Reply:
> Of course the fact that there is no mount for object makes it similar, from a pod's perspective, to a database.

Agreed.  Given that the data path is so different between file/block and object, it's better to describe it as such.  We've already seen that by lumping object in with file/block, there's been some assumptions about whether we're also proposing a POSIX analog for bucket interfaces, which we're not.